[Bug]: Replaced cron triggers retain retired evaluation state

### Bug type

Behavior bug (incorrect output/state without crash)

### Beta release blocker

No

### Summary

Replacing or removing a cron condition trigger preserves the retired trigger evaluation state, allowing the replacement automation to inherit stale deduplication state and silently never fire.

### Steps to reproduce

1. Create a cron automation with a condition trigger and persisted `triggerState`, `triggerEvalCount`, `lastTriggerEvalAtMs`, and `lastTriggerFireAtMs`.
2. Replace the condition script, remove the trigger, or transition the job to or from a script payload through either `cron.update` or declarative convergence.
3. Reload the persisted cron store.
4. Observe that the retired evaluation state remains attached to the new executable owner.

### Expected behavior

Trigger evaluation state is retained only while the exact condition trigger or script payload that owns it remains unchanged. Ownership transitions clear retired state unless replacement state is supplied explicitly.

### Actual behavior

All four state fields survive ordinary and declarative ownership transitions, so a replacement condition can receive the retired state and suppress a fire.

### OpenClaw version

Current `main` before the repair (`91b8a034535a1dab34a1cfe88b2caf52d1ee0c3a`).

### Operating system

macOS 26.6

### Install method

Source checkout with pnpm

### Model

N/A — deterministic cron persistence behavior

### Provider / routing chain

N/A — deterministic cron persistence behavior

### Additional provider/model setup details

N/A

### Logs

```shell
FAIL src/cron/service.declarative-jobs.test.ts > persists trigger-state ownership transitions and explicit replacements through ordinary update
FAIL src/cron/service.declarative-jobs.test.ts > persists trigger-state ownership transitions and explicit replacements through declarative convergence
AssertionError: condition replacement: triggerState: expected { owner: 'retired' } to deeply equal undefined
Tests 2 failed | 6 skipped
```

### Screenshots, recordings, and evidence

Focused persisted-store regression reproduces the failure through both mutation paths. No UI surface is involved.

### Impact and severity

Affected: Operators using condition-trigger or script-backed cron automations that are replaced, removed, or converted in place.

Severity: High because the replacement automation can silently produce no action.

Frequency: Deterministic whenever retired deduplication state affects the replacement executable.

Consequence: A configured automation appears valid but may never fire.

### Additional information

The persisted schema is sufficient. The repair belongs in the shared job mutation lifecycle and requires no schema bump or downstream guard.

--- a/src/cron/service.declarative-jobs.test.ts
+++ b/src/cron/service.declarative-jobs.test.ts
@@ -1,4 +1,5 @@
 import { describe, expect, it, vi } from "vitest";
+import { resolveCronJobConfigRevision } from "./config-revision.js";
 import { CronService } from "./service.js";
 import {
   createCronStoreHarness,
@@ -6,6 +7,7 @@ import {
   installCronTestHooks,
 } from "./service.test-harness.js";
 import type { CronAddResult } from "./service/state.js";
+import { loadCronStore } from "./store.js";
 import type { CronJob, CronJobCreate } from "./types.js";
 
 const logger = createNoopLogger();
@@ -44,6 +46,40 @@ function declarativeResult(result: CronAddResult) {
     throw new Error("expected declarative cron result");
   }
   return result;
+}
+
+const retiredTriggerState = {
+  triggerState: { owner: "retired" },
+  triggerEvalCount: 7,
+  lastTriggerEvalAtMs: 111,
+  lastTriggerFireAtMs: 222,
+  lastRunAtMs: 333,
+};
+
+type TriggerStateOwner = "condition" | "script" | "none";
+
+function ownedDeclaration(params: {
+  declarationKey: string;
+  owner: TriggerStateOwner;
+  script?: string;
+  state?: CronJobCreate["state"];
+  sameOwnerEdit?: boolean;
+}): CronJobCreate {
+  const script = params.script ?? 'return "original"';
+  return declaration({
+    declarationKey: params.declarationKey,
+    delivery: { mode: "none" },
+    trigger:
+      params.owner === "condition"
+        ? { script, ...(params.sameOwnerEdit ? { once: true } : {}) }
+        : undefined,
+    payload:
+      params.owner === "script"
+        ? { kind: "script", script, ...(params.sameOwnerEdit ? { timeoutSeconds: 45 } : {}) }
+        : { kind: "agentTurn", message: "report" },
+    ...(params.state ? { state: params.state } : {}),
+    ...(params.sameOwnerEdit && params.owner === "none" ? { displayName: "Updated report" } : {}),
+  });
 }
 
 describe("CronService declarative jobs", () => {
@@ -177,6 +213,134 @@ describe("CronService declarative jobs", () => {
     }
   });
 
+  it.each(["ordinary update", "declarative convergence"] as const)(
+    "persists trigger-state ownership transitions and explicit replacements through %s",
+    async (mutationPath) => {
+      const { storePath } = await makeStorePath();
+      const cron = createCronService(storePath);
+      await cron.start();
+      const cases: Array<{
+        name: string;
+        previous: TriggerStateOwner;
+        next: TriggerStateOwner;
+        replaceScript?: boolean;
+        sameOwnerEdit?: boolean;
+        replacementState?: CronJobCreate["state"];
+      }> = [
+        {
+          name: "condition replacement",
+          previous: "condition",
+          next: "condition",
+          replaceScript: true,
+        },
+        { name: "condition removal", previous: "condition", next: "none" },
+        { name: "condition to script", previous: "condition", next: "script" },
+        { name: "script replacement", previous: "script", next: "script", replaceScript: true },
+        { name: "script removal", previous: "script", next: "none" },
+        { name: "script to condition", previous: "script", next: "condition" },
+        { name: "no owner to condition", previous: "none", next: "condition" },
+        { name: "no owner to script", previous: "none", next: "script" },
+        {
+          name: "same condition owner",
+          previous: "condition",
+          next: "condition",
+          sameOwnerEdit: true,
+        },
+        { name: "same script owner", previous: "script", next: "script", sameOwnerEdit: true },
+        { name: "same absent owner", previous: "none", next: "none", sameOwnerEdit: true },
+        {
+          name: "explicit replacement state and count",
+          previous: "condition",
+          next: "condition",
+          replaceScript: true,
+          replacementState: { triggerState: null, triggerEvalCount: 0 },
+        },
+        {
+          name: "explicit replacement evaluation timestamps",
+          previous: "condition",
+          next: "script",
+          replacementState: { lastTriggerEvalAtMs: 444, lastTriggerFireAtMs: 555 },
+        },
+      ];
+      const persistedExpectations: Array<{ id: string; state: CronJob["state"]; name: string }> =
+        [];
+
+      try {
+        for (const [index, testCase] of cases.entries()) {
+          const declarationKey = `trigger-owner:${mutationPath}:${index}`;
+          const input = ownedDeclaration({
+            declarationKey,
+            owner: testCase.previous,
+            state: retiredTriggerState,
+          });
+          if (mutationPath === "ordinary update") {
+            delete input.declarationKey;
+          }
+          const result = await cron.add(input);
+          const created = "job" in result ? result.job : result;
+          expect(created.state, `${testCase.name}: create preserves explicit state`).toMatchObject(
+            retiredTriggerState,
+          );
+
+          const next = ownedDeclaration({
+            declarationKey,
+            owner: testCase.next,
+            script: testCase.replaceScript ? 'return "replacement"' : undefined,
+            state: testCase.replacementState,
+            sameOwnerEdit: testCase.sameOwnerEdit,
+          });
+          if (mutationPath === "ordinary update") {
+            await cron.update(created.id, {
+              trigger: next.trigger ?? null,
+              payload: next.payload,
+              displayName: next.displayName,
+              ...(testCase.replacementState ? { state: testCase.replacementState } : {}),
+            });
+          } else {
+            await cron.add(next);
+          }
+
+          const expectedState: CronJob["state"] = testCase.sameOwnerEdit
+            ? retiredTriggerState
+            : { lastRunAtMs: retiredTriggerState.lastRunAtMs, ...testCase.replacementState };
+          const persisted = (await loadCronStore(storePath)).jobs.find(
+            (entry) => entry.id === created.id,
+          );
+          expect(persisted?.state, `${testCase.name}: durable state`).toMatchObject(expectedState);
+          for (const field of [
+            "triggerState",
+            "triggerEvalCount",
+            "lastTriggerEvalAtMs",
+            "lastTriggerFireAtMs",
+          ] as const) {
+            expect(persisted?.state[field], `${testCase.name}: ${field}`).toEqual(
+              expectedState[field],
+            );
+          }
+          persistedExpectations.push({ id: created.id, state: expectedState, name: testCase.name });
+        }
+      } finally {
+        cron.stop();
+      }
+
+      const restarted = createCronService(storePath, false);
+      for (const expected of persistedExpectations) {
+        const persisted = await restarted.readJob(expected.id);
+        expect(persisted?.state, `${expected.name}: restart`).toMatchObject(expected.state);
+        for (const field of [
+          "triggerState",
+          "triggerEvalCount",
+          "lastTriggerEvalAtMs",
+          "lastTriggerFireAtMs",
+        ] as const) {
+          expect(persisted?.state[field], `${expected.name}: restart ${field}`).toEqual(
+            expected.state[field],
+          );
+        }
+      }
+    },
+  );
+
   it("checks update preconditions under the mutation lock", async () => {
     const { storePath } = await makeStorePath();
     const cron = createCronService(storePath);
@@ -192,6 +356,88 @@ describe("CronService declarative jobs", () => {
       expect(await cron.readJob(created.id)).toMatchObject({ displayName: "Daily report" });
     } finally {
       cron.stop();
+    }
+  });
+
+  it("rejects concurrent stale-owner updates across service instances sharing SQLite", async () => {
+    const { storePath } = await makeStorePath();
+    const first = createCronService(storePath);
+    const second = createCronService(storePath);
+    await first.start();
+    await second.start();
+
+    try {
+      const created = declarativeResult(
+        await first.add(
+          ownedDeclaration({
+            declarationKey: "trigger-owner:concurrent",
+            owner: "condition",
+            state: retiredTriggerState,
+          }),
+        ),
+      );
+      await second.readJob(created.id);
+      const staleRevision = resolveCronJobConfigRevision(created.job);
+      const replacementState = { triggerState: { owner: "replacement" }, triggerEvalCount: 23 };
+      const commitGuard = vi.fn();
+
+      await expect(
+        first.add(
+          ownedDeclaration({
+            declarationKey: "trigger-owner:concurrent",
+            owner: "condition",
+            script: 'return "invalid"',
+            state: { lastTriggerEvalAtMs: -1 },
+          }),
+          { commitGuard },
+        ),
+      ).rejects.toThrow("cron state.lastTriggerEvalAtMs must be a non-negative Date-valid integer");
+      expect(commitGuard).not.toHaveBeenCalled();
+      expect((await second.readJob(created.id))?.state).toMatchObject(retiredTriggerState);
+
+      const [replacement, stale] = await Promise.allSettled([
+        first.update(created.id, {
+          trigger: { script: 'return "replacement"' },
+          state: replacementState,
+        }),
+        second.updateWithPrecondition(
+          created.id,
+          {
+            displayName: "Stale owner",
+            trigger: { script: 'return "obsolete"' },
+            state: { triggerState: { owner: "obsolete" }, triggerEvalCount: 99 },
+          },
+          (current) => {
+            if (resolveCronJobConfigRevision(current) !== staleRevision) {
+              throw new Error("revision conflict");
+            }
+          },
+        ),
+      ]);
+
+      expect(replacement.status).toBe("fulfilled");
+      expect(stale).toMatchObject({ status: "rejected", reason: new Error("revision conflict") });
+      const persisted = await second.readJob(created.id);
+      expect(persisted?.state).toMatchObject(replacementState);
+      expect(persisted?.state.lastTriggerEvalAtMs).toBeUndefined();
+      expect(persisted?.state.lastTriggerFireAtMs).toBeUndefined();
+      expect(persisted?.displayName).toBe("Daily report");
+      expect(persisted?.trigger).toEqual({ script: 'return "replacement"' });
+
+      const currentRevision = resolveCronJobConfigRevision(persisted!);
+      await second.updateWithPrecondition(
+        created.id,
+        { displayName: "Current owner" },
+        (current) => {
+          if (resolveCronJobConfigRevision(current) !== currentRevision) {
+            throw new Error("revision conflict");
+          }
+        },
+      );
+      expect((await first.readJob(created.id))?.state).toMatchObject(replacementState);
+    } finally {
+      second.stop();
+      first.stop();
     }
   });
 

--- a/src/cron/service.trigger-eval.test.ts
+++ b/src/cron/service.trigger-eval.test.ts
@@ -72,10 +72,11 @@ async function runWhenDue(cron: CronService, jobId: string) {
 }
 
 describe("cron trigger evaluation", () => {
-  it("persists quiet evaluations without payload execution or run history", async () => {
-    const evaluateCronTrigger = vi.fn(async () => ({
+  it("persists quiet evaluations and fires replacement triggers with fresh state", async () => {
+    const replacementScript = 'return "replacement"';
+    const evaluateCronTrigger = vi.fn(async (params: Parameters<Evaluator>[0]) => ({
       kind: "evaluated" as const,
-      fire: false,
+      fire: params.script === replacementScript && params.state === undefined,
       state: { status: "green" },
     }));
     const harness = await createHarness({ evaluateCronTrigger });
@@ -113,6 +114,14 @@ describe("cron trigger evaluation", () => {
           jobId: job.id,
         }).entries,
       ).toEqual([]);
+
+      await harness.cron.update(job.id, { trigger: { script: replacementScript } });
+      expect(await runWhenDue(harness.cron, job.id)).toEqual({ ok: true, ran: true });
+      expect(evaluateCronTrigger).toHaveBeenLastCalledWith(
+        expect.objectContaining({ script: replacementScript, state: undefined }),
+      );
+      expect(harness.runIsolatedAgentJob).toHaveBeenCalledOnce();
+      expect(harness.cron.getJob(job.id)?.state.triggerEvalCount).toBe(1);
     } finally {
       harness.cron.stop();
     }
@@ -367,7 +376,7 @@ describe("cron trigger evaluation", () => {
       expect(stored?.trigger).toEqual(
         restore ? originalTrigger : { script: "replacement trigger", once: true },
       );
-      expect(stored?.state.triggerState).toEqual({ owner: "latest edit" });
+      expect(stored?.state.triggerState).toEqual(restore ? undefined : { owner: "latest edit" });
       expect(stored?.state.lastTriggerEvalAtMs).toBeUndefined();
       expect(stored?.state.nextRunAtMs).toEqual(expect.any(Number));
     } finally {
@@ -471,7 +480,11 @@ describe("cron trigger evaluation", () => {
         expect(stored?.payload).toMatchObject(
           restore ? originalPayload : { kind: "script", script: "return replacement" },
         );
-        expect(stored?.state.triggerState).toEqual({ owner: "current" });
+        expect(stored?.state.triggerState).toBeUndefined();
+        const persisted = (await loadCronStore(harness.storePath)).jobs.find(
+          (entry) => entry.id === job.id,
+        );
+        expect(persisted?.state.triggerState).toBeUndefined();
       } finally {
         completion.resolve({ status: "ok", stateChanged: true, state: { owner: "cleanup" } });
         harness.cron.stop();

--- a/src/cron/service/jobs.ts
+++ b/src/cron/service/jobs.ts
@@ -541,6 +541,7 @@ export function applyDeclarativeJobSpec(
   if (opts.enabledExplicit) {
     job.enabled = input.enabled;
   }
+  assertCronJobStateTimestamps(input.state ?? {});
   validateFullJob(
     job,
     {

--- a/src/cron/service/ops-mutations.ts
+++ b/src/cron/service/ops-mutations.ts
@@ -101,6 +101,7 @@ function finalizeUpdatedJob(params: {
   now: number;
   schedulingInputsRequested: boolean;
   scheduleChanged: boolean;
+  explicitTriggerState?: CronJobPatch["state"];
 }) {
   const { job, nextJob, now } = params;
   if (nextJob.schedule.kind === "every") {
@@ -136,6 +137,25 @@ function finalizeUpdatedJob(params: {
   // watcher. Equivalent resaves preserve it; disable/enable and source changes
   // rotate it in the same write that changes the public job definition.
   reconcileStreamSourceIdentity(job, nextJob);
+
+  const previousScript = job.payload.kind === "script" ? job.payload.script : undefined;
+  const nextScript = nextJob.payload.kind === "script" ? nextJob.payload.script : undefined;
+  if (job.trigger?.script !== nextJob.trigger?.script || previousScript !== nextScript) {
+    // Trigger and payload scripts share one durable state slot; only its exact
+    // executable owner may inherit it, while explicit replacement values win.
+    for (const field of [
+      "triggerState",
+      "triggerEvalCount",
+      "lastTriggerEvalAtMs",
+      "lastTriggerFireAtMs",
+    ] as const) {
+      if (params.explicitTriggerState && Object.hasOwn(params.explicitTriggerState, field)) {
+        Object.assign(nextJob.state, { [field]: params.explicitTriggerState[field] });
+      } else {
+        delete nextJob.state[field];
+      }
+    }
+  }
 
   // Only advance a recurring job's next run when the schedule/enabled inputs
   // actually changed. An idempotent re-save (same schedule, or re-enabling an
@@ -382,6 +402,7 @@ export async function add(
         now,
         schedulingInputsRequested: true,
         scheduleChanged: !isDeepStrictEqual(existing.schedule, nextJob.schedule),
+        explicitTriggerState: normalizedInput.state,
       });
       await persistUpdatedJob({ state, snapshot, previousJob: existing, nextJob });
       return { ...nextJob, created: false, updated: true, job: nextJob };
@@ -519,6 +540,7 @@ async function updateLoadedJob(params: {
       "trigger" in patch ||
       "pacing" in patch,
     scheduleChanged: patch.schedule !== undefined,
+    explicitTriggerState: patch.state,
   });
   const runtimeAuthorityMutation = consumeRuntimeAuthorityMutationOptions(opts);
   reconcileRuntimeAuthority({

